bluesky_widgets/qt/run_tree.py

```diff
@@ -0,0 +1,195 @@
+    def __init__(self, run):
+        self.run = run
+        self.children = []
+
+        uid = RunNode(self.run, 'uid', self.run.metadata['start']['uid'], None, self)
+        start = RunNode(self.run, 'start', 'dict', self.run.metadata['start'], self)
+        start.num_children = len(self.run.metadata['start'])
+        stop = RunNode(self.run, 'stop', 'dict', self.run.metadata['stop'], self)
+        stop.num_children = len(self.run.metadata['stop'])
+        streams = RunNode(self.run, 'streams', 'events (' + str(len(self.run)) + ')', None, self)
+        streams.num_children = len(self.run)
+        self.children = [uid, start, stop, streams]
+
+    def count(self):
+        return len(self.children)
+
+    def child(self, row):
+        if row >=0 and row < len(self.children):
+            return self.children[row]
+        return None
+
+
+class RunNode(object):
+    def __init__(self, run, key, value, data=None, parent=None):
+        self.run = run
+        self.parent = parent
+        self.key = key
+        self.value = value
+        self.data = data
+        self.children = []
+        self.num_children = 0
+
+    def child_count(self):
+        return self.num_children
+
+    def child(self, row):
+        """Lazily fill these in when we can..."""
+        if len(self.children) == 0 and self.num_children != len(self.children):
+            self.fill_children()
+
+        if row >= 0 and row < len(self.children):
+            return self.children[row]
+
+    def child_number(self):
+        return self.parent.children.index(self)
+
+    def fill_children(self):
+        """Handle special ones like streams first."""
+        if self.key == 'streams':
+            for stream in self.run:
+                num_events = self.run.metadata['stop']['num_events']
+                if stream in num_events:
+                    n = num_events[stream]
+                    if n == 0:
+                        value = '0 events'
+                    if n == 1:
+                        value = '1 event'
+                    else:
+                        value = str(n) + ' events'
+                else:
+                    value = 'null'
+                child = RunNode(self.run, stream, value, self.run[stream], self)
+                # Establish how many child nodes there will be.
+                child.num_children = len(self.run[stream].read().keys())
+                self.children.append(child)
+        elif self.data and isinstance(self.data, BlueskyEventStream):
+            # For now just display the keys in the stream.
+            node = RunNode(self.run, 'metadata', 'dict', self.data.metadata, self)
+            node.num_children = len(self.data.metadata)
+            self.children.append(node)
+            for key in self.data.read().keys():
+                child = RunNode(self.run, key, '', None, self)
+                self.children.append(child)
+        elif self.data and isinstance(self.data, dict):
```

Comment:
For `isinstance` checks on core types, prefer `collections.abc.Mapping` and `collections.abc.Iterable` because they will catch duck types as well as the specific types. A concrete example of where we care about this: we want to catch `list` and `tuple`.

Reply:
Done